core-api/src/main/java/org/silverpeas/core/calendar/CalendarEvent.java

```diff
@@ -0,0 +1,1314 @@
+
+  /**
+   * Recurs this event with the specified event recurrence.
+   *
+   * If the recurrence ends up at a given date or datetime, then this value is updated according to
+   * the period of time of this event:
+   *
+   * <ul>
+   *  <li>
+   *    The event is on all the day: the recurrence rule is updated to end at the given recurrence
+   *    ending date; the time part is removed.
+   *  </li>
+   *  <li>
+   *    The event starts and ends at a given datetime: the recurrence rule is updated to end at
+   *    a datetime with as date the given recurrence ending date and as time the time at which
+   *    this event usually starts.
+   *  </li>
+   * </ul>
+   * @param recurrence the recurrence defining the recurring property of this event.
+   * @return itself.
+   */
+  @Override
+  public CalendarEvent recur(final Recurrence recurrence) {
+    if (isOnAllDay() && recurrence.getFrequency().isHourly()) {
+      throw new IllegalArgumentException("Impossible to recur hourly an event on all day!");
+    }
+    this.recurrence = recurrence.startingAt(this.getStartDate());
+    return this;
+  }
+
+  @Override
+  public void unsetRecurrence() {
+    this.recurrence = Recurrence.NO_RECURRENCE;
+  }
+
+  /**
+   * Sets a description ot this event.
+   * @param description the description to set.
+   * @return itself.
+   */
+  public CalendarEvent withDescription(String description) {
+    setDescription(description);
+    return this;
+  }
+
+  /**
+   * Adds the specified attribute among the attributes of this event. Same as
+   * <pre>{@code
+   * getAttributes().add(attrName, attrValue);
+   * return this;}</pre>
+   * @param attrName the name of the attribute to add.
+   * @param attrValue the value of the attribute to add.
+   * @return itself.
+   */
+  public CalendarEvent withAttribute(String attrName, String attrValue) {
+    getAttributes().set(attrName, attrValue);
+    return this;
+  }
+
+  /**
+   * Adds the specified categories to the event. Same as
+   * <pre>{@code
+   * getCategories().addAll(categories);
+   * return this;}</pre>
+   * @param categories one or more categories with which this event will be categorized.
+   * @return itself.
+   */
+  public CalendarEvent withCategories(String ...categories) {
+    getCategories().addAll(categories);
+    return this;
+  }
+
+  /**
+   * Gets the recurrence of this recurring event. If the event isn't a recurring one, then returns
+   * NO_RECURRENCE.
+   * @return this event recurrence or NO_RECURRENCE.
+   */
+  @Override
+  public Recurrence getRecurrence() {
+    return this.recurrence;
+  }
+
+  /**
+   * Is this event occurring on all the day(s)?
+   * @return true if this event is occurring on all its day(s).
+   */
+  public boolean isOnAllDay() {
+    return getPeriod().isInDays();
+  }
+
+  /**
+   * BE CAREFUL, it is not possible to use the persistence when the ID is set by this method or
+   * {@link #setId(String)}.<br>
+   * TODO this behavior will be removed...
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1135)